iOS (App Extension): Friendly Support

dmiluski opened the pull request:
| Obligatory Gif |
| - |
| ![tenor](https://user-images.githubusercontent.com/5083390/77969800-78c89800-729f-11ea-9671-df89b0c04080.gif) |
| Extension |


## 🔧  Issue:
https://github.com/mapbox/mapbox-events-ios/issues/47#event-3180214494

## 🥅 Goal:
Play friendly with usage inside of an extension

## 📚  Context:
When using this module inside of an extension, developers will see errors in console:
'sharedApplication' is unavailable: not available on iOS (App Extension) - Use view controller based solutions where appropriate instead.

## 🗺 Approach:
1. Leverage view controller based solutions where appropriate (Available in iOS10 +
2. Inspect bundle to determine if if an extension or not to determine whether or not to access model information with Application or ExtensionContext

## 👷 Changes:
+ Add NSBundle Extensin to inspect whether in an Extension
+ Add NSExtensionContext to extract contentSizeScale from view controller based solution.
+ Inspect Bundle/Extension to construct Scale/State values

## 🔨 Unrelated Change:
+ Included DS_Store so it doesn’t clutter git diff

dmiluski committed:
iOS (App Extension): Friendly Support

Issue:
https://github.com/mapbox/mapbox-events-ios/issues/47#event-3180214494

Goal:
Play friendly with usage inside of an extension

Context:
When using this module inside of an extension, developers will see errors in console:
'sharedApplication' is unavailable: not available on iOS (App Extension) - Use view controller based solutions where appropriate instead.

Approach:
1. Leverage view controller based solutions where appropriate (Available in iOS10 +
2. Inspect bundle to determine if if an extension or not to determine whether or not to access model information with Application or ExtensionContext

Changes:
+ Add NSBundle Extensin to inspect whether in an Extension
+ Add NSExtensionContext to extract contentSizeScale from view controller based solution.
+ Inspect Bundle/Extension to construct Scale/State values

Unrelated Change:
+ Included DS_Store so it doesn’t clutter git diff

dmiluski commented on MapboxMobileEvents/Categories/UIKit+MMEMobileEvents.m:
```diff
@@ -82,3 +82,46 @@ - (NSInteger)mme_contentSizeScale {
+    NSInteger result = -9999;
+
+    if (@available(iOS 10, *)) {
+        NSString *sc = UIScreen.mainScreen.traitCollection.preferredContentSizeCategory;
```
Given UIApplication shared isn't available, grab preferredContentSizeCategory directly from UIScreen. Unfortunately this only became available in iOS10. Based on prior result of -9999 value, assumed this same default would be used in the event user is running < iOS10

dmiluski commented on MapboxMobileEvents/MMECommonEventData.m:
```diff
@@ -71,16 +72,22 @@ - (instancetype)init {
-            return MMEApplicationStateUnknown;
+
+    if (NSBundle.mme_isExtension) {
+        return MMEApplicationStateUnknown;
```
I wasn't quite sure how to port these values in the event of running in an extension. Extensions offer a much more limiting running time being killed immediately after dismissal.

Options:
- Return active by default if in extension?
- Return Unknown given there isn't a value to check?
- Other?

dmiluski replied:
Alternative:
Drop key/value all together if extension? Or the full event if extension?

rclee replied:
Let's add another value for `MMEApplicationStateExtension` and merge. We discussed this in a recent meeting and think it makes sense to let this fly with a known value to observe.

misterkerry reviewed (approved):
lgtm

rclee reviewed (approved):
Approved pending addition to ApplicationState. 

As an aside, how come we chose to fork here? I think Bitrise is having some trouble figuring out how to handle that...

dmiluski commented:
> Approved pending addition to ApplicationState.
> 
> As an aside, how come we chose to fork here? I think Bitrise is having some trouble figuring out how to handle that...

@rclee . Out of habit, before contributing to any repo I usually fork if option is available. This allows for contribution with or without collaborator status. Given this is a public repo, and others could also potentially contribute, this seemed okay at least for the time being? (I'll have to take a peak at bitrise and see if fork CI support is allowed)

A fork allows for a scratch pad to follow any organizational branch naming convention given any change I make doesn't clutter the main repo. In the past, if I > 10 branches I have felt that there may be a lack of focus on delivering features/value. I have seen 1,000+ branches in a repo and its usually a sign that team members aren't delivering features, or perhaps even worse, aren't collaborating to push work to the finish line and letting branches linger.

In doing so the main forked repo is allowed to remain focused and easy to discover what features are in the pipeline with long running feature branches such as Swift 3/4/5 support able to be done with focus, while features/bugfixes can be approached in the confines of a PR.

dmiluski commented:
@rclee https://github.com/mapbox/mapbox-events-ios/pull/204#discussion_r409245249

Will do

dmiluski committed:
Extension: Track App State

Problem:
Existing Application State is based on UIApplication which is unavailable in extensions.

Goal:
To gather potentially useful information that might be useful include additional state to bin extension events with extension state.

Changes:
+ Add Extension State
+ If extension, report state as extension

dmiluski commented:
There appears to be an issue with PRs from Forks. While investigating, will re-open from a local main repo branch.
